app/tests/src/db/test_bulk_ops.py

```diff
@@ -0,0 +1,89 @@
+"""Tests for bulk_ops module"""
+import operator
+import random
+from dataclasses import dataclass
+
+from psycopg import rows, sql
+
+import src.adapters.db as db
+from src.db import bulk_ops
+
+
+def test_bulk_upsert(db_session: db.Session):
+    conn = db_session.connection().connection
+    # Override mypy, because SQLAlchemy's DBAPICursor type doesn't specify the row_factory attribute, or that it functions as a context manager
+    with conn.cursor(row_factory=rows.class_row(Number)) as cur:  # type: ignore
```

Comment:
Would love to know if there's a better way of doing this. I also considered:
```
    db_client = db.PostgresDBClient()
    conn = db_client._engine.raw_connection()
```
but accessing `_engine` directly did not feel appropriate (and doesn't solve for the type issue in any case)

Reply:
hmm, not sure

Reply:
We could consider adding a `raw_connection()` method to the client class which does what you suggested. For the docs, mention that unless you're trying to do something very low level (ie. in psycopg) you'll almost never actually want to use it.

Reply:
I added this and included a comment with that context -- LMK what you think!